common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java

```diff
@@ -512,11 +510,82 @@ public void run() {
+     * @param request ClientCertRequest received from AzureActiveDirectoryWebViewClient.onReceivedClientCertRequest.
+     */
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    private void useSmartcardCertForAuth(@NonNull final X509Certificate cert,
+                                         @NonNull final char[] pin,
+                                         @NonNull final String slotAlias,
+                                         @NonNull final PivSession piv,
+                                         @NonNull final ClientCertRequest request) {
+        final String methodTag = TAG + "useSmartcardCertForAuth:";
+        //Need to add a PivProvider instance to the beginning of the array of Security providers in order for signature logic to occur.
+        //Note that this provider is removed when the UsbYubiKeyDevice connection is closed.
+        Security.insertProviderAt(new PivProvider(getPivProviderCallback()), 1);
+        try {
+            //Using KeyStore methods in order to generate PivPrivateKey.
+            //Loading null is needed for initialization.
+            KeyStore keyStore = KeyStore.getInstance("YKPiv", new PivProvider(piv));
+            keyStore.load(null);
+            //PivPrivateKey implements PrivateKey
+            PivPrivateKey pivPrivateKey = (PivPrivateKey) keyStore.getKey(slotAlias, pin);
+            //Cert chain only needs the cert to be used for authentication.
+            X509Certificate[] chain = new X509Certificate[]{cert};
+            //Dismiss the PIN dialog.
+            synchronized (sSmartcardDialogLock) {
+                mCurrentDialog.dismiss();
+                mCurrentDialog = null;
+            }
+            //Call proceed on ClientCertRequest with PivPrivateKey and cert chain.
+            request.proceed(pivPrivateKey, chain);
+        } catch (UnrecoverableKeyException | CertificateException | KeyStoreException | IOException | NoSuchAlgorithmException e) {
+            Logger.error(methodTag, e.getMessage(), e);
+            cancelCbaAndResetCurrentDialog();
+            //Show general error dialog.
+            showErrorDialog(R.string.smartcard_general_error_dialog_title, R.string.smartcard_general_error_dialog_message);
+        }
+    }
+
+    /**
+     * Used to provide PivProvider constructor a Callback that will establish a new PivSession when it is needed.
+     * @return A Callback which returns a Callback that will return a new PivSession instance.
+     */
+    private Callback<Callback<Result<PivSession, Exception>>> getPivProviderCallback() {
+        return new Callback<Callback<Result<PivSession, Exception>>>() {
+            @Override
+            public void invoke(@NonNull final Callback<Result<PivSession, Exception>> callback) {
+                synchronized (sDeviceLock) {
+                    mDevice.requestConnection(UsbSmartCardConnection.class, new Callback<Result<UsbSmartCardConnection, IOException>>() {
```

Comment:
mDevice could be unplugged during the process.

Reply:
added null check and invoke Result.failure on callback in such case, which throws an Exception with the message "Exception: Instance UsbYubiKitDevice variable (mDevice) is null."